File: Example_Objc/ViewController.m
//
//  Example
//  man
//
//  Created by man on 11/11/2018.
//  Copyright © 2018 man. All rights reserved.
//

#import "ViewController.h"
#import "AFURLSessionManager.h"
#import <WebKit/WebKit.h>
#import "CocoaDebugTool.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = @"Example";
    
    NSLog(@"hello world");
    NSLog(@"hello world red");
    NSLog(@"hello world yellow");
    NSLog(@"%d", 6666666);
    NSLog(@"unicode转换为中文");
    
    
    [self test_console_WKWebView];
    [self test_console_UIWebView];
    
    
    
    //Custom Messages
    [CocoaDebugTool logWithString:@"Custom Messages...."];
    [CocoaDebugTool logWithString:@"Custom Messages...." color:[UIColor redColor]];
    
    
    //save image
    [self saveImage:[UIImage imageNamed:@"111.png"]];
    
    //save txt
    [self saveTXT:@"hahahahahahahaha"];
    
    for (int i = 0; i < 20; i ++) {
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, i*2 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
            [self testHTTP];
        });
    }
}

- (void)saveImage:(UIImage *)image {
    NSString *imagePath = [NSHomeDirectory() stringByAppendingPathComponent:@"Documents/111"];
    [UIImagePNGRepresentation(image) writeToFile:imagePath atomically:YES];
}

- (void)saveTXT:(NSString *)txt {
    NSArray *documentArray = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *documentPath = [documentArray firstObject];
    NSLog(@"documentPath = %@",documentPath);
    NSString *filePath = [documentPath stringByAppendingPathComponent:@"ios.txt"];
    [[NSFileManager defaultManager] createFileAtPath:filePath contents:nil attributes:nil];
    
    NSString *iOSPath = [documentPath stringByAppendingPathComponent:@"ios.txt"];
    NSString *content = txt;
    [content writeToFile:iOSPath atomically:YES encoding:NSUTF8StringEncoding error:nil];
}

- (void)testHTTP {
    
    //1.AFNetworking
    NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
    AFURLSessionManager *manager = [[AFURLSessionManager alloc] initWithSessionConfiguration:configuration];
    NSURL *URL = [NSURL URLWithString:@"https://httpbin.org/get"];
    NSURLRequest *request = [NSURLRequest requestWithURL:URL];
    NSURLSessionDataTask *dataTask = [manager dataTaskWithRequest:request uploadProgress:^(NSProgress * _Nonnull uploadProgress) {
    } downloadProgress:^(NSProgress * _Nonnull downloadProgress) {
    } completionHandler:^(NSURLResponse * _Nonnull response, id  _Nullable responseObject, NSError * _Nullable error) {
        if (error) {
            NSLog(@"%@", error.localizedDescription);
        } else {
            NSLog(@"%@",response);
            NSLog(@"%@", responseObject);
        }
    }];
    [dataTask resume];
    
    //2.NSURLConnection
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        NSString *apiURLStr =[NSString stringWithFormat:@"https://httpbin.org/get"];
        NSMutableURLRequest *dataRqst = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:apiURLStr] cachePolicy:NSURLRequestUseProtocolCachePolicy timeoutInterval:30.0];
        NSHTTPURLResponse *response =[[NSHTTPURLResponse alloc] init];
        NSError *error = nil;
        NSData *responseData = [NSURLConnection sendSynchronousRequest:dataRqst returningResponse:&response error:&error];
        NSString *responseString = [[NSString alloc] initWithBytes:[responseData bytes] length:[responseData length] encoding:NSUTF8StringEncoding];
        if (error) {
            NSLog(@"%@", error.localizedDescription);
        }else{
            NSLog(@"%@", responseString);
        }
    });
    
    //3.NSURLSession
    NSMutableURLRequest *urlRequest = [[NSMutableURLRequest alloc] initWithURL:[NSURL URLWithString:@"https://httpbin.org/get"]];
    [urlRequest setHTTPMethod:@"GET"];
    NSURLSession *session = [NSURLSession sharedSession];
    NSURLSessionDataTask *dataTask_ = [session dataTaskWithRequest:urlRequest completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
        if(httpResponse.statusCode == 200) {
            NSError *parseError = nil;
            NSDictionary *responseDictionary = [NSJSONSerialization JSONObjectWithData:data options:0 error:&parseError];
            NSLog(@"%@",responseDictionary);
        }else{
            NSLog(@"%@", error.localizedDescription);
        }
    }];
    [dataTask_ resume];
}



- (void)test_console_WKWebView {
    WKWebView *webView = [WKWebView new];
    [self.view addSubview:webView];
    [webView loadHTMLString:[NSString stringWithContentsOfFile:[[NSBundle mainBundle] pathForResource:@"index" ofType:@"html"] encoding:NSUTF8StringEncoding error:nil] baseURL:[[NSBundle mainBundle] bundleURL]];
}

- (void)test_console_UIWebView {
    UIWebView *webView = [UIWebView new];
    [self.view addSubview:webView];
    [webView loadHTMLString:[NSString stringWithContentsOfFile:[[NSBundle mainBundle] pathForResource:@"index2" ofType:@"html"] encoding:NSUTF8StringEncoding error:nil] baseURL:[[NSBundle mainBundle] bundleURL]];
}

@end

File: Example_Swift/ViewController.swift
//
//  Example
//  man
//
//  Created by man on 11/11/2018.
//  Copyright © 2018 man. All rights reserved.
//

import UIKit
import WebKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        print("hello world")
        print("hello world red", color: .red)
        print("hello world blue", color: UIColor.red)
        print_UNICODE("unicode转换为中文")

        testHTTP()
        testRedirect()
        test_console_WKWebView()
        test_console_UIWebView()
    }
    
    
    func testHTTP() {
        
        //1.Alamofire
        request("https://httpbin.org/get").responseJSON { response in
            print(response, color: .red)
        }
        request("https://httpbin.org/post", method: .post, parameters: ["data": "Alamofire"], encoding: JSONEncoding.default).responseJSON { response in
            print(response, color: .red)
        }
        
        //2.SwiftHTTP
        HTTP.GET("https://httpbin.org/get") { response in
            print(response.text, color: .red)
        }
        HTTP.POST("https://httpbin.org/post", parameters: ["data": "SwiftHTTP"]) { response in
            print(response.text, color: .red)
        }
        HTTP.GET("https://www.baidu.com/img/bd_logo1.png") { response in
            print(response.description, color: .red)
        }
        HTTP.GET("https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png") { response in
            print(response.description, color: .red)
        }
        
        //3.Networking
        Networking(baseURL: "https://httpbin.org").get("/get") { (response, error) in
            print(response)
        }
        Networking(baseURL: "https://httpbin.org").post("/post", parameters: ["data": "Networking"]) { (response, error) in
            print(response)
        }
        
        //4.ASIHTTPRequest
        ASIHTTPRequestClient.shared().getJSON("https://httpbin.org/get", params: nil, successBlock: { (request, JSON) in
            print(JSON)
        }) { (request, error) in
            print(error?.localizedDescription)
        }
        ASIHTTPRequestClient.shared().postJSON("https://httpbin.org/post", params: ["data": "ASIHTTPRequest"], successBlock: { (request, JSON) in
            print(JSON)
        }) { (request, error) in
            print(error?.localizedDescription)
        }
        ASIHTTPRequestClient.shared().postForm("https://httpbin.org/post", params: ["data": "ASIHTTPRequest"], successBlock: { (request, JSON) in
            print(JSON)
        }) { (request, error) in
            print(error?.localizedDescription)
        }
        
        //5.large response data
//        HTTP.POST("http://api.ellabook.cn/rest/api/service", parameters: ["method": "ella.book.listAllPart"]) { response in
//            print(response.text, color: .red)
//        }
    }
    
    
    func testRedirect() {
        RedirectRequester.getRedirectInfo(with: "http://apple.com") { result, response in
            print("Redirect \(result): \(response)")
        }
    }
    
    
    
    func test_console_WKWebView() {
        let webView = WKWebView()
        self.view.addSubview(webView)
        do{
            let str = try String.init(contentsOfFile: Bundle.main.path(forResource: "index", ofType: "html") ?? "")
            webView.loadHTMLString(str, baseURL: Bundle.main.bundleURL)
        }catch{}
    }
    
    func test_console_UIWebView() {
        let webView = UIWebView()
        self.view.addSubview(webView)
        do{
            let str = try String.init(contentsOfFile: Bundle.main.path(forResource: "index2", ofType: "html") ?? "")
            webView.loadHTMLString(str, baseURL: Bundle.main.bundleURL)
        }catch{}
    }
}


fileprivate class RedirectRequester: NSObject, URLSessionTaskDelegate {
    
    private static var sharedInstance = RedirectRequester()
    private static var session: URLSession!
    
    func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse, newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void) {
        completionHandler(nil)
    }
    
    class func getRedirectInfo(with url: String, completionHandler: @escaping (Bool, String) -> Void) -> Void {
        if self.session == nil {
            let config = URLSessionConfiguration.default
            self.session = URLSession(configuration: config, delegate: RedirectRequester.sharedInstance, delegateQueue: nil)
        }
        
        let request = URLRequest(url: URL(string: url)!, cachePolicy: .reloadIgnoringLocalCacheData)
        
        let task = self.session.dataTask(with: request) { data, response, error in
            if let error = error {
                completionHandler(false, error.localizedDescription)
                return
            }
            
            if let httpResponse = response as? HTTPURLResponse {
                if (300...302).contains(httpResponse.statusCode) {
                    completionHandler(true, httpResponse.allHeaderFields["Location"] as! String)
                    return
                }
            }
            
            completionHandler(false, "Redirect request failed")
        }
        task.resume()
    }
}

